hcloud/resource_hcloud_reversedns.go

```diff
@@ -0,0 +1,241 @@
+package hcloud
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hetznercloud/hcloud-go/hcloud"
+)
+
+func resourceReverseDNS() *schema.Resource {
+	return &schema.Resource{
+		Read:   resourceReverseDNSRead,
+		Create: resourceReverseDNSCreate,
+		Delete: resourceReverseDNSDelete,
+
+		Schema: map[string]*schema.Schema{
+			"server_id": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"floating_ip_id": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"ip_address": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"dns_ptr": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceReverseDNSRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*hcloud.Client)
+	ctx := context.Background()
+
+	ipAddress := string(d.Get("ip_address").(string))
+	rdnsType := ""
+	id, ok := d.GetOk("server_id")
+	if ok == false {
+		id, ok = d.GetOk("floating_ip_id")
+		switch {
+		case ok == false:
+			log.Printf("[WARN] Invalid id (%s), removing from state: %v", d.Id(), ok)
+			d.SetId("")
+			return nil
+		case ok == true:
+			rdnsType = "floating_ip"
+		}
+	} else {
+		rdnsType = "server"
+	}
+	ip := net.ParseIP(d.Get("ip_address").(string))
+	if ip == nil {
+		log.Printf("[WARN] The ip you provide (%s) is not valid, removing from state", ipAddress)
+		d.SetId("")
+		return nil
+	}
+	switch {
+	case rdnsType == "floating_ip":
+		floatingIP, _, err := client.FloatingIP.GetByID(ctx, id.(int))
+		if err != nil {
+			return err
+		}
+		if floatingIP == nil {
+			log.Printf("[WARN] Floating IP (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		switch floatingIP.Type {
+		case "ipv4":
+			if strings.Count(ipAddress, ".") != 3 {
+				log.Printf("[WARN] Floating IP (%s) is an ipv4 floating ip but you dont write a valid one, removing from state", d.Id())
+				d.SetId("")
+				return nil
+			}
+		case "ipv6":
+			if strings.Count(ipAddress, ":") > 1 {
+				log.Printf("[WARN] Floating IP (%s) is an ipv6 but you write an invalid ip, removing from state", d.Id())
+				d.SetId("")
+				return nil
+			}
+		}
+		d.Set("dns_ptr", floatingIP.DNSPtrForIP(ip))
+		d.SetId(fmt.Sprintf("f-%d-"+ipAddress, id))
```

Comment:
Please create a new function that generates the id for floating ips and servers.

Something like this:
```go
func generateRDNSID(server *hcloud.Server, floatingIP *hcloud.FloatingIP, ip string) string {
	if server != nil {
		return fmt.Sprintf("s-%d-%s", id, ip)
	}
	if floatingIP != nil {
		return fmt.Sprintf("f-%d-%s", id, ip)
	}
	return ""
}
```
Also note that if you use `fmt.Sprintf` you should use it with multiple parameters and not mix it with string concatenation.